@wdio/allure-reporter cucumber step reporter adds blank steps before and after each scenario bug fixing

JordiSAGE opened the pull request:
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)
PR for fixing the bug: https://github.com/webdriverio/webdriverio/issues/10358

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

### Reviewers: @webdriverio/project-committers

JordiSAGE commented:
> @JordiSAGE All Cucumber hooks have empty title, thats why you have empty steps in report. If you exclude all empty steps from report users get lost all information attached to such steps(images, params, link, etc). It seems unacteaptable

This is just limited to the hook steps with empty title, but in that case, what do you suggest in order to fix the issue? Maybe to limit the lines 544 and 578 when (!disableMochaHooks)?
Thanks for the suggestion!

JordiSAGE commented:
> > @JordiSAGE All Cucumber hooks have empty title, thats why you have empty steps in report. If you exclude all empty steps from report users get lost all information attached to such steps(images, params, link, etc). It seems unacteaptable
> 
> @BorisOsipov This is just limited to the hook steps with empty title on the onHookStart and onHookEnd events, but in that case, what do you suggest in order to fix the issue? Maybe to limit the lines 544 and 578 when (!disableMochaHooks)? Thanks for the suggestion!

BorisOsipov commented:
@JordiSAGE 

Reporting hooks in Allure can be challenging. I have spent a few days trying to make it work effectively within the scope of https://github.com/webdriverio/webdriverio/issues/9928, but without success.

I would suggest keeping the hooks reported. However, if the hook.title is empty and it is a cucumber framework with useCucumberStepReporter set to true, I recommend forcing the hook title to be set as "Unnamed cucumber hook."

I understand that it may not be the ideal solution and that it can result in "empty" hooks for some users. However, this approach allows for the inclusion of custom steps within hooks and attachments, providing more flexibility and functionality.

dstoutnutrien commented:
@BorisOsipov Could there be a `disableCucumberHooks` option which defaults to false, so users who don't want those empty steps can remove them explicitly?

BorisOsipov commented:
@dstoutnutrien 

Yes, it is a great option. However, imagine a scenario where you don't care about failed hooks because you don't use them. But suddenly, you end up using them and have the disableCucumberHooks option enabled. This could result in false positive report results and introduce bugs into production.

What is the best solution: keeping useless hooks in the report forever or missing out on bugs in the system under test due to an ugly report?

Another solution is for contributors to thoroughly check all possible cases when a before/after hook fails with different artifacts. It is important to ensure that none of these cases are missed, especially when a failed hook has no effect on reporting a scenario as failed.

BorisOsipov commented:
I think we should always have an excellent report for scenarios in which tests fail, for the sake of instances when they are successful. Even if they have some unsightly empty steps.

rsivasubramaniam commented:
@JordiSAGE @BorisOsipov 

Hello Fellas,

Having used + customized the Java version of Allure -- My 2 cents below as the **Java + Allure + Cuke** version is much more mature than the **WDIO + Allure + Cuke** version

1. Yes the

BorisOsipov commented:
@dstoutnutrien

>In what case might reporting that blank step/hook be helpful?

In all cases :) It may contain test related data 
![image](https://github.com/webdriverio/webdriverio/assets/6514276/7bbe9ea8-5749-44d6-af57-1f8cb96139d0)

I've checked with current fix tests aren't run at all with error smth about missing current step\test. So it is definitially breaking change.

Also in case of failed After hook test marked as `Passed` - it is ridiculous.
![image](https://github.com/webdriverio/webdriverio/assets/6514276/f40c42d8-e1c4-45ce-811f-18052d754e21)

JordiSAGE commented:
As @dstoutnutrien mentioned, I can't see the case in which these blank lines would be helpful, even the onHookEnd already includes the test fail creation in case of error, and honestly, the lines 544 and 578 seem to be added by mistake.

JordiSAGE commented:
@BorisOsipov Following your suggestions, I've reverted the previous changes, but alternatively I've added the disableMochaHooks checking sentence, that in case that this is false, it will not show the hook as passed, but still keeping it in case of error, I hope it will match your expectations.
Thanks for your review!

![image](https://github.com/webdriverio/webdriverio/assets/33693547/6a8919a1-d810-4786-945e-d2a206c800a3)

![image](https://github.com/webdriverio/webdriverio/assets/33693547/a1418104-4af2-4eaf-8d13-fb484887a5fa)

christian-bromann commented:
@BorisOsipov good to go?

BorisOsipov commented:
@christian-bromann unfortunately, I am not sure. I will flook into this tomorrow.

BorisOsipov commented:
@JordiSAGE 

> alternatively I've added the disableMochaHooks checking sentence, that in case that this is false, it will not show the hook as passed (when disableMochaHooks is set to true), but still keeping it in case of error, I hope it will match your expectations.


I am not happy with such an approach. First of all, disableMochaHooks  should not affect any cases unrelated to Mocha integration. As a Cucumber user, such an approach is non-obvious.

Secondly, I still insist on reporting all hooks in Cucumber to avoid cases where custom steps in hooks are not being reported. To achieve this, the best approach is to report all hooks. Due to a limitation in Cucumber, we cannot obtain the hook names. Therefore, if the `hook.title` is empty and it is a cucumber framework with `useCucumberStepReporte`r set to true, I recommend forcing the hook title to be set as "Unnamed cucumber hook" or simply "Unnamed hook". It will allow reporting all hooks, even if their "body" is empty and seemingly useless (although they may contain some user steps, attachments, parameters, etc.).

I understand that such an approach may make the report a bit ugly and verbose for you, but for the common case, it is appropriate because we need to cover all user cases.

I want to doubly emphasize that I am aware the report may look a bit ugly, but we have to maintain backward compatibility.

BorisOsipov reviewed (changes requested):
See comment in PR

JordiSAGE commented:
Hi again @BorisOsipov and @christian-bromann 

It can be a good solution, putting the message "Unnamed cucumber hook" or "Unnamed hook", in case to keep coherence in the code and to avoid cases where custom steps in hooks are not being reported, as you mentioned, but this is not clear functionally speaking, since the end users will see a blank space or the messages you proposed, those would be completely irrelevant as well as confusing for them, hope you understand, other than that, again from my point of view, this should have been discussed as part of the bug itself, or when it has been reported, even rejecting it as a bug, thus avoiding for having to provide the different solutions, again I hope that makes sense, so at this point I should insist on removing the blank passed steps (with or without messages), as you mentioned when `useCucumberStepReporter` is set to true, the `hook.title` is empty, and additionally by checking that the hook doesn't contain any steps, attachments, parameters, etc.
For doing so, I've removed the previous changes, and made the required change to only show the onTestStart step when the title is not empty or it contains any attachment and exclusively for when the `useCucumberStepReporter` variable is set to true, otherwiser keeping the steps as before.

Thank you again for your support and patience!

BorisOsipov commented:
>t I should insist on removing the blank passed steps (with or without messages), as you mentioned when useCucumberStepReporter is set to true, the hook.title is empty, and additionally by checking that the hook doesn't contain any steps, attachments, parameters, etc.

Nice. Just do it :)

>For doing so, I've removed the previous changes, and made the required change to only show the onTestStart step when the title is not empty or it contains any attachment and exclusively for when the useCucumberStepReporter variable is set to true, otherwiser keeping the steps as before.

These changes aren't effective. You simply moved the check for hook.title from one function to another. Additionally, the datatable for hooks is always empty. But the worst part is that this code still leads to corrupted reports. 
You should never allow cases where you don't call startStep/Test(testTitle) but call endStep/Test() 
and vice versa. It leads to a broken test-step-substep hierarchy and a broken report.

It seems that I haven't explained the problem very clearly. Let me do it.
There is a repository with an example available at https://github.com/BorisOsipov/wdio-allure-10475

In this case, we have before and after hooks where I used the Allure API to enrich the report with additional data.
https://github.com/BorisOsipov/wdio-allure-10475/blob/master/steps.ts#L21-L40
https://github.com/BorisOsipov/wdio-allure-10475/blob/master/steps.ts#L43-L49

Additionally, I have some hooks where I don't use the Allure API.
https://github.com/BorisOsipov/wdio-allure-10475/blob/master/steps.ts#L43-L49
https://github.com/BorisOsipov/wdio-allure-10475/blob/master/wdio.conf.ts#L56-L58

Currently, all tests have passed, and I have a correct but visually unappealing report (it contains steps with empty names).
But all my data, steps, attachments are correct placed as I wanted.
See screen: https://imgur.com/a/ni2akyA

If I make any changes to the reporter where we remove "empty" hooks from the report, I will have a partially correct report where all empty steps are gone. However, all my custom Allure steps, attachments, and other elements will be mixed into the test body, creating a chaotic and confusing presentation.

See screen: https://imgur.com/a/ZTbwtnq

So the goal of this bug fix should be to remove all empty steps from the report, but not to remove steps where I have my own created attachments, sub-steps, etc. Or even better, give the "empty" named steps some names like "Unnamed cucumber hook" or "Unnamed hook" to help users expand the step and see the sub-steps and data.

I hope it is now clear why we can't simply remove hooks where the hook.title is empty.

JordiSAGE commented:
Hi @BorisOsipov 

First, apologies for the mess in the code, I can see your point, and I'm doing my best to understand all the different cases for providing the best code I could as requested in the bug, but as shown I still don't have your experience to be able to do it.
I'm going to base my coming changes on the resources you gave me and provide the best possible solution.
Stay tuned!

Thank you again for your suggestions and resources!

BorisOsipov commented:
Hi @JordiSAGE 

Yes that's good👍
Could you add a unit test `wdio-allure-reporter/tests/cucumber.suite.test.ts` to ensure we don't run into any regression?

JordiSAGE commented:
Hi again @BorisOsipov 

I've added the corresponding mocha tests on cucumber test file to both test empty hooks and hooks with dependant steps attached.
Hope this coverage will be enough, fell free to let me know!

Thank you in advance.

BorisOsipov commented:
Hi @JordiSAGE Yes, these tests are sufficient.

Please, have a look into failed unit tests. The current implementation seems to be flawed
https://github.com/webdriverio/webdriverio/actions/runs/5380296095/jobs/9765373726?pr=10475#step:7:527

JordiSAGE commented:
Hello @BorisOsipov

I've fixed the issues when running the cucumber steps by modifying the code provided on onHookEnd method, apart from improve the mocha test cases, but there is still an issue on https://github.com/webdriverio/webdriverio/actions/runs/5380296095/jobs/9765373726?pr=10475#step:7:576
I can't manage to get rid of the Unhandled exception, more than to put a .catch() method on the promise, what makes the step as passed, as follows:
`it('should add failed custom step', async () => {
    await AllureReporter.step('custom step', () => {
        return Promise.reject(new Error('an error here')).catch(function () {
            console.log("Promise Rejected");
       })
    })

    const [, call] = vi.mocked(process.emit).mock.calls[0] as any
    expect(process.emit).toHaveBeenCalledTimes(1)
    expect(call.steps).toHaveLength(1)
    expect(call.steps[0]).toMatchObject({
        name: 'custom step',
        status: Status.PASSED,
        stage: Stage.FINISHED,
    })
})`

This part has been written by @christian-bromann some time ago, so maybe he can help on it, otherwise I don't know what to do to keep it as exception without throwing the unhandled error.

Thank you again!

christian-bromann reviewed (approved):
It seems like you were able to resolve that issue. The code looks good to me 👍 

@BorisOsipov should we merge?

BorisOsipov commented on packages/wdio-allure-reporter/tests/cucumber.suite.test.ts:
```diff
@@ -189,9 +189,9 @@ describe('reporter option "useCucumberStepReporter" set to true', () => {
+            expect(allureResult.steps).toHaveLength(2)
             expect(allureResult.steps[1].attachments).toHaveLength(1)
-            expect(allureResult.steps[1].attachments[0].name).toEqual('Console Logs')
+            expect(allureResult.steps[1].attachments[0].name).toEqual('Page source')
```
```suggestion
            expect(allureResult.steps[0].attachments[0].name).toEqual('Console Logs')
```

BorisOsipov commented on packages/wdio-allure-reporter/tests/cucumber.suite.test.ts:
```diff
@@ -189,9 +189,9 @@ describe('reporter option "useCucumberStepReporter" set to true', () => {
         it('should have the console log add', () => {
-            expect(allureResult.steps).toHaveLength(3)
+            expect(allureResult.steps).toHaveLength(2)
             expect(allureResult.steps[1].attachments).toHaveLength(1)
```
```suggestion
            expect(allureResult.steps[0].attachments).toHaveLength(1)
```

BorisOsipov reviewed (approved):
LGTM

BorisOsipov commented:
>@BorisOsipov should we merge?

Yes, I fixed test a bit and now CI jobs fail on bidi generation. IDK how to fix it.

christian-bromann commented:
> IDK how to fix it.

I am on it and will merge this PR as soon as we have a green build again.

christian-bromann commented:
Great job @JordiSAGE 👏 thanks a lot!

JordiSAGE commented:
Thank you @BorisOsipov and @christian-bromann for your help, appreciate it!
